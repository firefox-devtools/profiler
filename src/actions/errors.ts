/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// This file contains actions related to error handling.

import { Action } from 'firefox-profiler/types';

export function fatalError(error: Error): Action {
  return {
    type: 'FATAL_ERROR',
    error,
  };
}
