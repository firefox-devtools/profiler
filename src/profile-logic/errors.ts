/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import type { RequestedLib } from 'firefox-profiler/types';

// Used during the symbolication process to express that we couldn't find
// symbols for a specific library
export class SymbolsNotFoundError extends Error {
  library: RequestedLib;
  errors: Error[];

  constructor(message: string, library: RequestedLib, ...errors: Error[]) {
    super(
      [message, ...errors.map((e) => ` - ${e.name}: ${e.message}`)].join('\n')
    );
    // Workaround for a babel issue when extending Errors
    (this as any).__proto__ = SymbolsNotFoundError.prototype;
    this.name = 'SymbolsNotFoundError';
    this.library = library;
    this.errors = errors;
  }
}
