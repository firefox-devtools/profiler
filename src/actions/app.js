/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import { oneLine } from 'common-tags';
import {
  getSelectedTab,
  getDataSource,
  getLocalTrackOrderByPid,
} from 'firefox-profiler/selectors/url-state';
import {
  getTrackThreadHeights,
  getIsEventDelayTracksEnabled,
  getIsExperimentalCPUGraphsEnabled,
  getIsExperimentalProcessCPUTracksEnabled,
} from 'firefox-profiler/selectors/app';
import {
  getLocalTracksByPid,
  getThreads,
  getCounters,
} from 'firefox-profiler/selectors/profile';
import { sendAnalytics } from 'firefox-profiler/utils/analytics';
import {
  stateFromLocation,
  withHistoryReplaceStateSync,
} from 'firefox-profiler/app-logic/url-handling';
import { finalizeProfileView } from './receive-profile';
import { fatalError } from './errors';
import {
  addEventDelayTracksForThreads,
  initializeLocalTrackOrderByPid,
  addProcessCPUTracksForProcess,
} from 'firefox-profiler/profile-logic/tracks';
import { selectedThreadSelectors } from 'firefox-profiler/selectors/per-thread';
import {
  getIsCPUUtilizationProvided,
  getAreThereAnyProcessCPUCounters,
} from 'firefox-profiler/selectors/cpu';

import type {
  Profile,
  ThreadsKey,
  CssPixels,
  Action,
  ThunkAction,
  UrlState,
  UploadedProfileInformation,
  IndexIntoCategoryList,
} from 'firefox-profiler/types';
import type { TabSlug } from 'firefox-profiler/app-logic/tabs-handling';
import type {
  BrowserConnection,
  BrowserConnectionStatus,
} from 'firefox-profiler/app-logic/browser-connection';

export function changeSelectedTab(selectedTab: TabSlug): ThunkAction<void> {
  return (dispatch, getState) => {
    const previousTab = getSelectedTab(getState());
    if (previousTab !== selectedTab) {
      sendAnalytics({
        hitType: 'pageview',
        page: selectedTab,
      });
      dispatch({
        type: 'CHANGE_SELECTED_TAB',
        selectedTab,
      });
    }
  };
}

export function changeProfilesToCompare(profiles: string[]): Action {
  return {
    type: 'CHANGE_PROFILES_TO_COMPARE',
    profiles,
  };
}

export function startFetchingProfiles(): Action {
  return { type: 'START_FETCHING_PROFILES' };
}

export function urlSetupDone(): ThunkAction<void> {
  return (dispatch, getState) => {
    dispatch({ type: 'URL_SETUP_DONE' });

    // After the url setup is done, we can successfully query our state about its
    // initial page.
    const dataSource = getDataSource(getState());
    sendAnalytics({
      hitType: 'pageview',
      page: dataSource === 'none' ? 'home' : getSelectedTab(getState()),
    });
    sendAnalytics({
      hitType: 'event',
      eventCategory: 'datasource',
      eventAction: dataSource,
    });
  };
}

export function show404(url: string): Action {
  return { type: 'ROUTE_NOT_FOUND', url };
}

export function changeSidebarOpenState(tab: TabSlug, isOpen: boolean): Action {
  return { type: 'CHANGE_SIDEBAR_OPEN_STATE', tab, isOpen };
}

export function invalidatePanelLayout(): Action {
  return { type: 'INCREMENT_PANEL_LAYOUT_GENERATION' };
}

/**
 * The viewport component provides a hint to use shift to zoom scroll. The first
 * time a user does this, the hint goes away.
 */
export function setHasZoomedViaMousewheel() {
  return { type: 'HAS_ZOOMED_VIA_MOUSEWHEEL' };
}

/**
 * This function is called when we start setting up the initial url state.
 * It takes the location and profile data, converts the location into url
 * state and then dispatches relevant actions to finalize the view.
 * `profile` parameter can be null when the data source can't provide the profile
 * and the url upgrader step is not needed (e.g. 'from-browser').
 */
export function setupInitialUrlState(
  location: Location,
  profile: Profile | null,
  browserConnection: BrowserConnection | null
): ThunkAction<void> {
  return (dispatch) => {
    let urlState;
    try {
      urlState = stateFromLocation(location, profile);
    } catch (e) {
      if (e.name === 'UrlUpgradeError') {
        // The error is an URL upgrade error, let's fire a fatal error.
        // If there's a service worker update, the class `ServiceWorkerManager`
        // will automatically reload in case the new code knows how to handle
        // this URL version.
        dispatch(fatalError(e));
        return;
      }
      // The location could not be parsed, show a 404 instead.
      console.error(e);
      dispatch(show404(location.pathname + location.search));
      return;
    }

    // Validate the initial URL state. We can't refresh on a from-file URL.
    if (urlState.dataSource === 'from-file') {
      urlState = null;
    }

    // Normally having multiple dispatches is an anti pattern, but here it's
    // necessary because we are doing different things inside those actions and
    // they can't be merged because we are also calling those seperately on
    // other parts of the code.
    // The first dispatch here updates the url state, then changes state as the url
    // setup is done, and lastly finalizes the profile view since everything is set up now.
    // All of this is done while the history is replaced, as this is part of the initial
    // load process.
    withHistoryReplaceStateSync(() => {
      dispatch(updateUrlState(urlState));
      dispatch(finalizeProfileView(browserConnection));
      dispatch(urlSetupDone());
    });
  };
}

/**
 * This function is called when a browser navigation event happens. A new UrlState
 * is generated when the window.location is serialized, or the state is pulled out of
 * the history API.
 */
export function updateUrlState(newUrlState: UrlState | null): Action {
  return { type: 'UPDATE_URL_STATE', newUrlState };
}

export function reportTrackThreadHeight(
  threadsKey: ThreadsKey,
  height: CssPixels
): ThunkAction<void> {
  return (dispatch, getState) => {
    const trackThreadHeights = getTrackThreadHeights(getState());
    const previousHeight = trackThreadHeights[threadsKey];
    if (previousHeight !== height) {
      // Guard against unnecessary dispatches. This could happen frequently.
      dispatch({
        type: 'UPDATE_TRACK_THREAD_HEIGHT',
        height,
        threadsKey,
      });
    }
  };
}

/**
 * This action dismisses the newly published state. This happens when a user first
 * uploads a profile. We only want to remember this when we fist open the profile.
 */
export function dismissNewlyPublished(): Action {
  return { type: 'DISMISS_NEWLY_PUBLISHED' };
}

/**
 * Called when a user has started dragging a file. Used for loading
 * profiles with the drag and drop component.
 */
export function startDragging(): Action {
  return { type: 'START_DRAGGING' };
}

/**
 * Called when a user has stopped dragging a file.
 */
export function stopDragging(): Action {
  return { type: 'STOP_DRAGGING' };
}

/**
 * Called when a custom drag and drop overlay is mounted. This lets
 * the app know that we shouldn't create a default overlay.
 */
export function registerDragAndDropOverlay(): Action {
  return { type: 'REGISTER_DRAG_AND_DROP_OVERLAY' };
}

/**
 * Called when a custom drag and drop overlay is unmounted.
 */
export function unregisterDragAndDropOverlay(): Action {
  return { type: 'UNREGISTER_DRAG_AND_DROP_OVERLAY' };
}

/*
 * This action enables the event delay tracks. They are hidden by default because
 * they are usually for power users and not so meaningful for average users.
 * There is no UI that triggers this action in the profiler interface. Instead,
 * users have to enable this from the developer console by writing this line:
 * `experimental.enableEventDelayTracks()`
 */
export function enableEventDelayTracks(): ThunkAction<boolean> {
  return (dispatch, getState) => {
    if (getIsEventDelayTracksEnabled(getState())) {
      console.error(
        'Tried to enable the event delay tracks, but they are already enabled.'
      );
      return false;
    }

    if (
      selectedThreadSelectors.getSamplesTable(getState()).eventDelay ===
      undefined
    ) {
      // Return early if the profile doesn't have eventDelay values.
      console.error(oneLine`
        Tried to enable the event delay tracks, but this profile does
        not have eventDelay values. It is likely an older profile.
      `);
      return false;
    }

    const oldLocalTracks = getLocalTracksByPid(getState());
    const localTracksByPid = addEventDelayTracksForThreads(
      getThreads(getState()),
      oldLocalTracks
    );
    const localTrackOrderByPid = initializeLocalTrackOrderByPid(
      getLocalTrackOrderByPid(getState()),
      localTracksByPid,
      null
    );
    dispatch({
      type: 'ENABLE_EVENT_DELAY_TRACKS',
      localTracksByPid,
      localTrackOrderByPid,
    });

    return true;
  };
}

/*
 * This action enables the CPU graph tracks. They are hidden by default because
 * they are usually for power users and not so meaningful for average users.
 * There is no UI that triggers this action in the profiler interface. Instead,
 * users have to enable this from the developer console by writing this line:
 * `experimental.enableCPUGraphs()`
 */
export function enableExperimentalCPUGraphs(): ThunkAction<boolean> {
  return (dispatch, getState) => {
    if (getIsExperimentalCPUGraphsEnabled(getState())) {
      console.error(
        'Tried to enable the CPU graph tracks, but they are already enabled.'
      );
      return false;
    }

    if (!getIsCPUUtilizationProvided(getState())) {
      // Return early if the profile doesn't have threadCPUDelta values.
      console.error(oneLine`
        Tried to enable the CPU graph tracks, but this profile does
        not have threadCPUDelta values. It is likely an older profile.
      `);
      return false;
    }

    dispatch({
      type: 'ENABLE_EXPERIMENTAL_CPU_GRAPHS',
    });

    return true;
  };
}

/*
 * This action enables the process CPU tracks. They are hidden by default because
 * the front-end work is not done for this data yet.
 * There is no UI that triggers this action in the profiler interface. Instead,
 * users have to enable this from the developer console by writing this line:
 * `experimental.enableExperimentalProcessCPUTracks()`
 */
export function enableExperimentalProcessCPUTracks(): ThunkAction<boolean> {
  return (dispatch, getState) => {
    if (getIsExperimentalProcessCPUTracksEnabled(getState())) {
      console.error(
        'Tried to enable the process CPU tracks, but they are already enabled.'
      );
      return false;
    }

    if (
      !getIsCPUUtilizationProvided(getState()) &&
      getAreThereAnyProcessCPUCounters(getState())
    ) {
      // Return early if the profile doesn't have threadCPUDelta values or
      // doesn't have any experimental process CPU counters.
      console.error(oneLine`
        Tried to enable the process CPU tracks, but this profile does
        not have threadCPUDelta values or process CPU threads.
      `);
      return false;
    }

    const oldLocalTracks = getLocalTracksByPid(getState());
    const localTracksByPid = addProcessCPUTracksForProcess(
      getCounters(getState()),
      oldLocalTracks
    );
    const localTrackOrderByPid = initializeLocalTrackOrderByPid(
      getLocalTrackOrderByPid(getState()),
      localTracksByPid,
      null
    );

    dispatch({
      type: 'ENABLE_EXPERIMENTAL_PROCESS_CPU_TRACKS',
      localTracksByPid,
      localTrackOrderByPid,
    });

    return true;
  };
}

/**
 * This caches the profile data in the local state for synchronous access.
 */
export function setCurrentProfileUploadedInformation(
  uploadedProfileInformation: UploadedProfileInformation | null
): Action {
  return {
    type: 'SET_CURRENT_PROFILE_UPLOADED_INFORMATION',
    uploadedProfileInformation,
  };
}

export function profileRemotelyDeleted(): Action {
  // Ideally we should store the current profile data in a local indexeddb, and
  // set the URL to /local/<indexeddb-key>.
  return { type: 'PROFILE_REMOTELY_DELETED' };
}

export function updateBrowserConnectionStatus(
  browserConnectionStatus: BrowserConnectionStatus
): Action {
  return { type: 'UPDATE_BROWSER_CONNECTION_STATUS', browserConnectionStatus };
}

export function toggleOpenCategoryInSidebar(
  kind: string,
  category: IndexIntoCategoryList
): Action {
  return {
    type: 'TOGGLE_SIDEBAR_OPEN_CATEGORY',
    kind,
    category,
  };
}
