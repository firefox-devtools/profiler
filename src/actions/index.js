/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import * as app from './app';
import * as icons from './icons';
import * as profileView from './profile-view';
import * as publish from './publish';
import * as receiveProfile from './receive-profile';
import * as zippedProfiles from './zipped-profiles';
import * as code from './code';

export default Object.assign(
  {},
  app,
  icons,
  profileView,
  publish,
  receiveProfile,
  zippedProfiles,
  code
);
