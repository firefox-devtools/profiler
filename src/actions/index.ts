/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
import * as app from './app';
import * as icons from './icons';
import * as l10n from './l10n';
import * as profileView from './profile-view';
import * as publish from './publish';
import * as receiveProfile from './receive-profile';
import * as zippedProfiles from './zipped-profiles';
import * as code from './code';

export default Object.assign(
  {},
  app,
  icons,
  l10n,
  profileView,
  publish,
  receiveProfile,
  zippedProfiles,
  code
);
