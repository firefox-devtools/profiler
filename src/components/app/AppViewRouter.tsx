/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import { PureComponent } from 'react';
import explicitConnect from 'firefox-profiler/utils/connect';

import { ProfileViewer } from './ProfileViewer';
import { ZipFileViewer } from './ZipFileViewer';
import { Home } from './Home';
import { CompareHome } from './CompareHome';
import { ProfileRootMessage } from './ProfileRootMessage';
import { getView } from 'firefox-profiler/selectors/app';
import { getHasZipFile } from 'firefox-profiler/selectors/zipped-profiles';
import {
  getDataSource,
  getProfilesToCompare,
} from 'firefox-profiler/selectors/url-state';
import { ProfileLoaderAnimation } from './ProfileLoaderAnimation';
import { UploadedRecordingsHome } from './UploadedRecordingsHome';
import { assertExhaustiveCheck } from 'firefox-profiler/utils/flow';

import type { AppViewState, State, DataSource } from 'firefox-profiler/types';

import type { ConnectedProps } from 'firefox-profiler/utils/connect';
import { Localized } from '@fluent/react';

const ERROR_MESSAGES_L10N_ID: { [key: string]: string } = Object.freeze({
  'from-browser': 'AppViewRouter--error-unpublished',
  'from-post-message': 'AppViewRouter--error-from-post-message',
  unpublished: 'AppViewRouter--error-unpublished',
  'from-file': 'AppViewRouter--error-from-file',
  local: 'AppViewRouter--error-local',
  public: 'AppViewRouter--error-public',
  'from-url': 'AppViewRouter--error-from-url',
  compare: 'AppViewRouter--error-compare',
});

type AppViewRouterStateProps = {
  readonly view: AppViewState;
  readonly dataSource: DataSource;
  readonly profilesToCompare: string[] | null;
  readonly hasZipFile: boolean;
};

type AppViewRouterProps = ConnectedProps<{}, AppViewRouterStateProps, {}>;

class AppViewRouterImpl extends PureComponent<AppViewRouterProps> {
  override render() {
    const { view, dataSource, profilesToCompare, hasZipFile } = this.props;
    const phase = view.phase;

    // We're using a switch to assert that all values for the dataSource has
    // been checked. This is useful when we add a new dataSource, as Flow will
    // error here if we forget to update this code.
    switch (dataSource) {
      case 'none':
        return <Home />;
      case 'compare':
        if (profilesToCompare === null) {
          return <CompareHome />;
        }
        break;
      case 'uploaded-recordings':
        return <UploadedRecordingsHome />;
      case 'from-browser':
      case 'from-post-message':
      case 'unpublished':
      case 'from-file':
      case 'local':
      case 'public':
      case 'from-url':
        break;
      default:
        throw assertExhaustiveCheck(dataSource);
    }

    switch (phase) {
      case 'INITIALIZING':
      case 'PROFILE_LOADED':
      case 'DATA_RELOAD':
        return <ProfileLoaderAnimation />;
      case 'FATAL_ERROR': {
        let message =
          ERROR_MESSAGES_L10N_ID[dataSource] || 'AppViewRouter--error-public';
        let additionalMessage = null;
        if (view.error) {
          if (view.error.name === 'SafariLocalhostHTTPLoadError') {
            message = 'AppViewRouter--error-from-localhost-url-safari';
          } else {
            console.error(view.error);
            additionalMessage = (
              <>
                <p>{view.error.toString()}</p>
                <p>The full stack has been written to the Web Console.</p>
              </>
            );
          }
        }

        return (
          <Localized
            id={message}
            attrs={{ title: true }}
            elems={{
              // WebKit bug link, only used for AppViewRouter--message-from-localhost-url-safari
              a: (
                <a
                  href="https://bugs.webkit.org/show_bug.cgi?id=171934"
                  target="_blank"
                  rel="noopener noreferrer"
                />
              ),
            }}
          >
            <ProfileRootMessage
              additionalMessage={additionalMessage}
              showLoader={false}
              showBackHomeLink={true}
            >{`missing translation for ${message}`}</ProfileRootMessage>
          </Localized>
        );
      }
      case 'DATA_LOADED':
        // The data is now loaded. This could be either a single profile, or a zip file
        // with multiple profiles. Only show the ZipFileViewer if the data loaded is a
        // Zip file, and there is no stored path into the zip file.
        return hasZipFile ? <ZipFileViewer /> : <ProfileViewer />;
      case 'TRANSITIONING_FROM_STALE_PROFILE':
        return null;
      case 'ROUTE_NOT_FOUND':
      default:
        // Assert with TypeScript that we've handled all the cases, as the only thing left
        // should be 'ROUTE_NOT_FOUND' or 'PROFILE_LOADED'.
        phase as 'ROUTE_NOT_FOUND';
        return (
          <Localized
            id="AppViewRouter--route-not-found--home"
            attrs={{ specialMessage: true }}
          >
            <Home specialMessage="The URL you tried to reach was not recognized." />
          </Localized>
        );
    }
  }
}

export const AppViewRouter = explicitConnect<{}, AppViewRouterStateProps, {}>({
  mapStateToProps: (state: State) => ({
    view: getView(state),
    dataSource: getDataSource(state),
    profilesToCompare: getProfilesToCompare(state),
    hasZipFile: getHasZipFile(state),
  }),
  component: AppViewRouterImpl,
});
