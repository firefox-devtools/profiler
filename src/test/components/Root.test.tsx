/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// We want to test these components in isolation and tightly control the actions
// dispatched and avoid any side-effects.  That's why we mock this module and
// return dummy thunk actions that return a Promise.
jest.mock('../../actions/receive-profile', () => ({
  // These mocks will get their implementation in the `setup` function.
  // Otherwise the implementation is wiped before the test starts.
  // See https://github.com/facebook/jest/issues/7573 for more info.
  retrieveProfileFromBrowser: jest.fn(),
  retrieveProfileFromStore: jest.fn(),
  retrieveProfilesToCompare: jest.fn(),
}));

// We mock <ProfileViewer> because it's complex and we should really test it
// elsewhere. Using a custom name makes it possible to test that  _this_
// component is used.
jest.mock('../../components/app/ProfileViewer', () => ({
  ProfileViewer: 'profile-viewer',
}));
// We mock <Home> as well because it brings too much noise in snapshots and it's
// overly tested in another test file.
jest.mock('../../components/app/Home', () => ({
  Home: 'home',
}));
jest.mock('../../components/app/CompareHome', () => ({
  CompareHome: 'compare-home',
}));
// ListOfPublishedProfiles depends on IDB and renders asynchronously, so we'll
// just test we want to render it, but otherwise test it more fully in a
// separate test file.
jest.mock('../../components/app/ListOfPublishedProfiles', () => ({
  ListOfPublishedProfiles: 'list-of-published-profiles',
}));
import { Provider } from 'react-redux';

import { render, act } from 'firefox-profiler/test/fixtures/testing-library';
import { AppViewRouter } from '../../components/app/AppViewRouter';
import { ProfileLoader } from '../../components/app/ProfileLoader';
import { updateUrlState, changeProfilesToCompare } from '../../actions/app';
import { fatalError } from '../../actions/errors';

// Because this module is mocked but we want the real actions in the test, we
// use `jest.requireActual` here.
const {
  temporaryError,
  viewProfile,
  waitingForProfileFromBrowser,
  waitingForProfileFromStore,
} = jest.requireActual('../../actions/receive-profile');
// These functions are mocks
import {
  retrieveProfileFromStore,
  retrieveProfileFromBrowser,
  retrieveProfilesToCompare,
} from '../../actions/receive-profile';
import { stateFromLocation } from '../../app-logic/url-handling';
import { TemporaryError } from '../../utils/errors';

import { blankStore } from '../fixtures/stores';
import { getProfileFromTextSamples } from '../fixtures/profiles/processed-profile';
import type { Action } from 'firefox-profiler/types';

describe('app/AppViewRouter', function () {
  it('renders an initial home', function () {
    const { container } = setup();

    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders the addon loading page, and the profile view after capturing', function () {
    const { container, dispatch, navigateToFromBrowserProfileLoadingPage } =
      setup();

    navigateToFromBrowserProfileLoadingPage();
    dispatch(waitingForProfileFromBrowser());
    expect(container.firstChild).toMatchSnapshot();
    // We do not check that retrieveProfileFromBrowser gets called, because the call
    // happens asynchronously.

    const { profile } = getProfileFromTextSamples(`A`);
    dispatch(viewProfile(profile));
    expect(container.firstChild).toMatchSnapshot();
  });

  it('does not try to retrieve a profile when moving from from-browser to public', function () {
    const {
      container,
      dispatch,
      navigateToFromBrowserProfileLoadingPage,
      navigateToStoreLoadingPage,
    } = setup();

    navigateToFromBrowserProfileLoadingPage();
    dispatch(waitingForProfileFromBrowser());
    const { profile } = getProfileFromTextSamples(`A`);
    dispatch(viewProfile(profile));

    navigateToStoreLoadingPage();
    expect(container.firstChild).toMatchSnapshot();
    expect(retrieveProfileFromStore).not.toHaveBeenCalled();
  });

  it('renders the profile view', function () {
    const { container, dispatch, navigateToStoreLoadingPage } = setup();

    navigateToStoreLoadingPage();
    dispatch(waitingForProfileFromStore());
    expect(container.firstChild).toMatchSnapshot();
    expect(retrieveProfileFromStore).toHaveBeenCalledWith('ThisIsAFakeHash');

    const { profile } = getProfileFromTextSamples(`A`);
    dispatch(viewProfile(profile));
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders temporary errors', function () {
    const { container, dispatch, navigateToStoreLoadingPage } = setup();

    navigateToStoreLoadingPage();
    dispatch(
      temporaryError(
        new TemporaryError('This is a temporary error, wait a bit more.', {
          count: 3,
          total: 10,
        })
      )
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders an home when the user pressed back after an error', function () {
    const {
      container,
      dispatch,
      navigateToFromFileProfileLoadingPage,
      navigateBackToHome,
    } = setup();

    navigateToFromFileProfileLoadingPage();
    dispatch(fatalError(new Error('Error while loading profile')));
    expect(container.firstChild).toMatchSnapshot();
    expect(console.error).toHaveBeenCalled();

    navigateBackToHome();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('does not render back home link when not opened with from-file', function () {
    const {
      container,
      dispatch,
      navigateToStoreLoadingPage,
    } = setup();

    navigateToStoreLoadingPage();
    dispatch(fatalError(new Error('Error while loading profile')));
    expect(container.firstChild).toMatchSnapshot();
    expect(console.error).toHaveBeenCalled();
  });

  it('renders a compare home when navigating to /compare/, then loads when changing profiles', () => {
    const { container, dispatch, navigateToCompareHome } = setup();

    navigateToCompareHome();
    expect(container.querySelector('compare-home')).toBeTruthy();

    const url1 = 'http://fake-url.com/hash/1';
    const url2 = 'http://fake-url.com/hash/2';
    dispatch(changeProfilesToCompare([url1, url2]));

    expect(container.querySelector('.loading')).toBeTruthy();
    expect(retrieveProfilesToCompare).toHaveBeenCalledWith([url1, url2]);
  });

  it('renders a uploaded-recordings page when navigating to /uploaded-recordings', () => {
    const { container, navigateToMyProfiles } = setup();
    navigateToMyProfiles();
    expect(container.querySelector('list-of-published-profiles')).toBeTruthy();
  });
});

function setup() {
  // Let's silence the error output to the console
  jest.spyOn(console, 'error').mockImplementation(() => {});
  // Flow doesn't know these actions are jest mocks.
  (retrieveProfileFromBrowser as any).mockImplementation(() => async () => {});
  (retrieveProfileFromStore as any).mockImplementation(() => async () => {});
  (retrieveProfilesToCompare as any).mockImplementation(() => async () => {});

  const store = blankStore();
  const renderResult = render(
    <Provider store={store}>
      <>
        <ProfileLoader />
        <AppViewRouter />
      </>
    </Provider>
  );

  function actAndDispatch(what: Action) {
    act(() => {
      store.dispatch(what);
    });
  }

  function navigateToStoreLoadingPage() {
    const newUrlState = stateFromLocation({
      pathname: '/public/ThisIsAFakeHash/calltree',
      search: '',
      hash: '',
    });
    actAndDispatch(updateUrlState(newUrlState));
  }

  function navigateToFromBrowserProfileLoadingPage() {
    const newUrlState = stateFromLocation({
      pathname: '/from-browser/',
      search: '',
      hash: '',
    });
    actAndDispatch(updateUrlState(newUrlState));
  }

  function navigateToFromFileProfileLoadingPage() {
    const newUrlState = stateFromLocation({
      pathname: '/from-file/calltree',
      search: '',
      hash: '',
    });
    actAndDispatch(updateUrlState(newUrlState));
  }

  function navigateBackToHome() {
    const newUrlState = stateFromLocation({
      pathname: '/',
      hash: '',
      search: '',
    });
    actAndDispatch(updateUrlState(newUrlState));
  }

  function navigateToCompareHome() {
    const newUrlState = stateFromLocation({
      pathname: '/compare/',
      hash: '',
      search: '',
    });
    actAndDispatch(updateUrlState(newUrlState));
  }

  function navigateToMyProfiles() {
    const newUrlState = stateFromLocation({
      pathname: '/uploaded-recordings/',
      hash: '',
      search: '',
    });
    actAndDispatch(updateUrlState(newUrlState));
  }

  return {
    ...renderResult,
    dispatch: actAndDispatch,
    navigateToStoreLoadingPage,
    navigateToFromBrowserProfileLoadingPage,
    navigateToFromFileProfileLoadingPage,
    navigateBackToHome,
    navigateToCompareHome,
    navigateToMyProfiles,
  };
}
