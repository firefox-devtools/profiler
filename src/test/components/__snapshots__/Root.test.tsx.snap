// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`app/AppViewRouter does not try to retrieve a profile when moving from from-browser to public 1`] = `<profile-viewer />`;

exports[`app/AppViewRouter renders an home when the user pressed back after an error 1`] = `
<div
  class="rootMessageContainer"
>
  <div
    class="rootMessage"
  >
    <h1
      class="rootMessageTitle"
    >
      Firefox Profiler
    </h1>
    <div
      class="rootMessageText"
    >
      <p>
        Could not download the profile.
      </p>
    </div>
    <div
      class="rootMessageAdditional"
    >
      <p>
        Error: Error while loading profile
      </p>
      <p>
        The full stack has been written to the Web Console.
      </p>
    </div>
    <div
      class="rootBackHomeLink"
    >
      <a
        href="/"
      >
        Back to home
      </a>
    </div>
  </div>
</div>
`;

exports[`app/AppViewRouter renders an home when the user pressed back after an error 2`] = `<home />`;

exports[`app/AppViewRouter renders an initial home 1`] = `<home />`;

exports[`app/AppViewRouter renders temporary errors 1`] = `
<div
  class="rootMessageContainer"
>
  <div
    class="rootMessage"
  >
    <h1
      class="rootMessageTitle"
    >
      Firefox Profiler
    </h1>
    <div
      class="rootMessageText"
    >
      <p>
        Downloading and processing the profile…
      </p>
    </div>
    <div
      class="rootMessageAdditional"
    >
      <p>
        This is a temporary error, wait a bit more.
      </p>
      <p>
        Tried 3 times out of 10.
      </p>
    </div>
    <div
      class="rootBackHomeLink"
    >
      <a
        href="/"
      >
        Back to home
      </a>
    </div>
    <div
      class="loading"
    >
      <div
        class="loading-div loading-div-1 loading-row-1"
      />
      <div
        class="loading-div loading-div-2 loading-row-2"
      />
      <div
        class="loading-div loading-div-3 loading-row-3"
      />
      <div
        class="loading-div loading-div-4 loading-row-3"
      />
      <div
        class="loading-div loading-div-5 loading-row-4"
      />
      <div
        class="loading-div loading-div-6 loading-row-4"
      />
      <div
        class="loading-div loading-div-7 loading-row-4"
      />
      <div
        class="loading-div loading-div-8 loading-row-4"
      />
      <div
        class="loading-div loading-div-9 loading-row-4"
      />
      <div
        class="loading-div loading-div-10 loading-row-4"
      />
    </div>
  </div>
</div>
`;

exports[`app/AppViewRouter renders the addon loading page, and the profile view after capturing 1`] = `
<div
  class="rootMessageContainer"
>
  <div
    class="rootMessage"
  >
    <h1
      class="rootMessageTitle"
    >
      Firefox Profiler
    </h1>
    <div
      class="rootMessageText"
    >
      <p>
        Importing the profile directly from ⁨Firefox⁩…
      </p>
    </div>
    <div
      class="loading"
    >
      <div
        class="loading-div loading-div-1 loading-row-1"
      />
      <div
        class="loading-div loading-div-2 loading-row-2"
      />
      <div
        class="loading-div loading-div-3 loading-row-3"
      />
      <div
        class="loading-div loading-div-4 loading-row-3"
      />
      <div
        class="loading-div loading-div-5 loading-row-4"
      />
      <div
        class="loading-div loading-div-6 loading-row-4"
      />
      <div
        class="loading-div loading-div-7 loading-row-4"
      />
      <div
        class="loading-div loading-div-8 loading-row-4"
      />
      <div
        class="loading-div loading-div-9 loading-row-4"
      />
      <div
        class="loading-div loading-div-10 loading-row-4"
      />
    </div>
  </div>
</div>
`;

exports[`app/AppViewRouter renders the addon loading page, and the profile view after capturing 2`] = `<profile-viewer />`;

exports[`app/AppViewRouter renders the profile view 1`] = `
<div
  class="rootMessageContainer"
>
  <div
    class="rootMessage"
  >
    <h1
      class="rootMessageTitle"
    >
      Firefox Profiler
    </h1>
    <div
      class="rootMessageText"
    >
      <p>
        Downloading and processing the profile…
      </p>
    </div>
    <div
      class="loading"
    >
      <div
        class="loading-div loading-div-1 loading-row-1"
      />
      <div
        class="loading-div loading-div-2 loading-row-2"
      />
      <div
        class="loading-div loading-div-3 loading-row-3"
      />
      <div
        class="loading-div loading-div-4 loading-row-3"
      />
      <div
        class="loading-div loading-div-5 loading-row-4"
      />
      <div
        class="loading-div loading-div-6 loading-row-4"
      />
      <div
        class="loading-div loading-div-7 loading-row-4"
      />
      <div
        class="loading-div loading-div-8 loading-row-4"
      />
      <div
        class="loading-div loading-div-9 loading-row-4"
      />
      <div
        class="loading-div loading-div-10 loading-row-4"
      />
    </div>
  </div>
</div>
`;

exports[`app/AppViewRouter renders the profile view 2`] = `<profile-viewer />`;
